examples/control.py

```diff
@@ -0,0 +1,227 @@
+#
+LqrSpec = collections.namedtuple('LqrSpec', 'Q q R r M A B')
+
+
+dot = np.dot
+mm = np.matmul
+
+
+def mv(mat, vec):
+  assert mat.ndim == 2
+  assert vec.ndim == 1
+  return dot(mat, vec)
+
+
+LOOP_VIA_SCAN = False
+
+
+def fori_loop(lo, hi, loop, init):
+  if LOOP_VIA_SCAN:
+    return scan_fori_loop(lo, hi, loop, init)
+  else:
+    return lax.fori_loop(lo, hi, loop, init)
+
+
+def scan_fori_loop(lo, hi, loop, init):
+  def scan_f(x, t):
+    return loop(t, x), ()
+  x, _ = lax.scan(scan_f, init, np.arange(lo, hi))
+  return x
+
+
+def trajectory(dynamics, U, x0):
+  '''Unrolls `X[t+1] = dynamics(t, X[t], U[t])`, where `X[0] = x0`.'''
+  T, _ = U.shape
+  d, = x0.shape
+
+  X = np.zeros((T + 1, d))
+  X = jo.index_update(X, jo.index[0], x0)
+
+  def loop(t, X):
+    x = dynamics(t, X[t], U[t])
+    X = jo.index_update(X, jo.index[t + 1], x)
+    return X
+
+  return fori_loop(0, T, loop, X)
+
+
+def make_lqr_approx(p):
+  T = p.horizon
+
+  def approx_timestep(t, x, u):
+    M = jacfwd(grad(p.cost, argnums=2), argnums=1)(t, x, u).T
+    Q = jacfwd(grad(p.cost, argnums=1), argnums=1)(t, x, u)
+    R = jacfwd(grad(p.cost, argnums=2), argnums=2)(t, x, u)
+    q, r = grad(p.cost, argnums=(1, 2))(t, x, u)
+    A, B = jacobian(p.dynamics, argnums=(1, 2))(t, x, u)
+    return Q, q, R, r, M, A, B
+
+  _approx = vmap(approx_timestep)
+
+  def approx(X, U):
+    assert X.shape[0] == T + 1 and U.shape[0] == T
+    U_pad = np.vstack((U, np.zeros((1,) + U.shape[1:])))
+    Q, q, R, r, M, A, B = _approx(np.arange(T + 1), X, U_pad)
+    return LqrSpec(Q, q, R[:T], r[:T], M[:T], A[:T], B[:T])
+
+  return approx
+
+
+def lqr_solve(spec):
+  EPS = 1e-7
+  T, control_dim, _ = spec.R.shape
+  _, state_dim, _ = spec.Q.shape
+
+  K = np.zeros((T, control_dim, state_dim))
+  k = np.zeros((T, control_dim))
+
+  def rev_loop(t_, state):
+    t = T - t_ - 1
+    spec, P, p, K, k = state
+
+    Q, q = spec.Q[t], spec.q[t]
+    R, r = spec.R[t], spec.r[t]
+    M = spec.M[t]
+    A, B = spec.A[t], spec.B[t]
+
+    AtP = mm(A.T, P)
+    BtP = mm(B.T, P)
+    G = R + mm(BtP, B)
+    H = mm(BtP, A) + M.T
+    h = r + mv(B.T, p)
+    K_ = -np.linalg.solve(G + EPS * np.eye(G.shape[0]), H)
+    k_ = -np.linalg.solve(G + EPS * np.eye(G.shape[0]), h)
+    P_ = Q + mm(AtP, A) + mm(K_.T, H)
+    p_ = q + mv(A.T, p) + mv(K_.T, h)
+
+    K = jo.index_update(K, jo.index[t], K_)
+    k = jo.index_update(k, jo.index[t], k_)
+    return spec, P_, p_, K, k
+
+  _, P, p, K, k = fori_loop(
+      0, T, rev_loop,
+      (spec, spec.Q[T + 1], spec.q[T + 1], K, k))
+
+  return K, k
+
+
+def lqr_predict(spec, x0):
+  T, control_dim, _ = spec.R.shape
+  _, state_dim, _ = spec.Q.shape
+
+  K, k = lqr_solve(spec)
+
+  def fwd_loop(t, state):
+    spec, X, U = state
+    A, B = spec.A[t], spec.B[t]
+    u = mv(K[t], X[t]) + k[t]
+    x = mv(A, X[t]) + mv(B, u)
+    X = jo.index_update(X, jo.index[t + 1], x)
+    U = jo.index_update(U, jo.index[t], u)
```

Comment:
We could potentially avoid the `index_update` if we wrote this in terms of `scan`, just because `scan` does the index-update for us. But I'm not sure if that'd actually end up simpler.

Not actionable, just thought I'd mention the idea!

Reply:
Agreed. Good point!